cmake/GzPython.cmake

```diff
@@ -24,3 +24,101 @@ find_package(Python3 ${GZ_PYTHON_VERSION} QUIET)
 if(Python3_EXECUTABLE AND NOT PYTHON_EXECUTABLE)
   set(PYTHON_EXECUTABLE ${Python3_EXECUTABLE})
 endif()
+
+# Macro Definitions #
+
+#################################################
+# Insert options required to control the include of pybind11
+#
+# Usage:
+#
+#     gz_add_pybind_project_settings()
+#
+# Variables defined by this macro:
+#
+#  SKIP_PYBIND11                                Skip generating of Python bindings if OFF
+#  USE_SYSTEM_PATHS_FOR_PYTHON_INSTALLATION     Install modules in default system path if ON
+#  USE_DIST_PACKAGES_FOR_PYTHON                 Use dist-package over site-package for module installation if ON
+#
+macro(gz_add_pybind_project_settings)
+  set(skip_pybind11_default_value OFF)
+  if (MSVC)
+    # We're disabling pybind11 by default on Windows because they
+    # don't have active CI on them for now.
+    set(skip_pybind11_default_value ON)
+  endif()
+
+  option(SKIP_PYBIND11
+        "Skip generating Python bindings via pybind11"
+        ${skip_pybind11_default_value})
+
+  include(CMakeDependentOption)
+
+  cmake_dependent_option(USE_SYSTEM_PATHS_FOR_PYTHON_INSTALLATION
+        "Install Python modules in standard system paths in the system"
+        OFF "NOT SKIP_PYBIND11" OFF)
+
+  cmake_dependent_option(USE_DIST_PACKAGES_FOR_PYTHON
+        "Use dist-packages instead of site-package to install Python modules"
+        OFF "NOT SKIP_PYBIND11" OFF)
+endmacro()
+
+
+#################################################
+# Search for pybind11 and its dependencies if SKIP_PYBIND11 is not set
+#
+# Usage:
+#
+#     gz_add_pybind_project_settings() <- required!
+#     gz_search_for_pybind()
+#
+# Variables defined by this macro:
+#
+#  Python3_*            Python result variables from FindPython3 module
+#  pybind11_*           pybind11 result variables from pybind11 module
+#
+macro(gz_search_for_pybind)
+  if (SKIP_PYBIND11)
+    message(STATUS "SKIP_PYBIND11 set - disabling python bindings")
+  else()
+    find_package(Python3 ${GZ_PYTHON_VERSION} QUIET COMPONENTS Interpreter Development)
```

Comment:
https://github.com/gazebosim/gz-math/pull/588 updated the code in gz-math to use `OPTIONAL_COMPONENTS` here.
https://github.com/gazebosim/gz-math/blob/ef2211c6292929ea6a8471815cf83315c1b891cb/CMakeLists.txt#L102-L104

Mind updating this to match?

cc @scpeters

Reply:
I think I haven't been completely consistent with using `COMPONENTS` or `OPTIONAL_COMPONENTS`

Reply:
ok, in gz-math8 and sdformat15, I used made finding the Python3 Interpreter `REQUIRED` and used `OPTIONAL_COMPONENTS` to find Development, while we didn't apply this logic in gz-transport or gz-sim, probably because I didn't realize the inconsistency

* https://github.com/gazebosim/gz-math/blob/main/CMakeLists.txt#L102-L105
* https://github.com/gazebosim/sdformat/blob/sdf14/CMakeLists.txt#L87-L90
* https://github.com/gazebosim/gz-transport/blob/main/CMakeLists.txt#L61
* https://github.com/gazebosim/gz-sim/blob/main/CMakeLists.txt#L209

I think I do recommend the approach used in gz-math and sdformat, but I also recognize the lack of flexibility in finding packages from a macro. I would suggest calling `find_package(Python3 ...)` from within the macro if `NOT Python3_Development_FOUND`, which would allow a package to use their own call to `find_package(Python3)` but fall back to a good default if it hasn't been searched for yet

Reply:
Implemented as suggested in `GzPybind.cmake` with `OPTIONAL_COMPONENTS` as Python might be used for other tasks where the `Development` component is not required.